src/rcc.rs

```diff
@@ -1,26 +1,341 @@
             apb1: APB1(()),
             apb2: APB2(()),
             bdcr: BDCR(()),
             csr: CSR(()),
-            cfgr: CFGR { hclk: None, pclk1: None, pclk2: None, sysclk: None }
+            cfgr: CFGR {
+                hclk: None,
+                pclk1: None,
+                pclk2: None,
+                sysclk: clocking::SysClkSource::MSI(
+                    clocking::MediumSpeedInternalRC::new(4_000_000, false)
+                    ),
+            }
         }
     }
 }
 
+pub mod clocking {
+    //! The `clocking` module contains representations of the various objects in the STM32L4x6
+    //! clock tree (see Reference Manual Figs 15 and 16) useful for wiring them up.
+    //!
+    //! There are two main concepts: sources (enum variants) and clocks (structs). For example, the
+    //! SYSCLK clock can be driven by any one of four sources: HSI, MSI, HSI, and PLLCLK. This
+    //! knowledge is encoded in the SysClkSources enum.
+    //!
+    //! Each enum variant contains information about the clock it represents. Some clocks are
+    //! configurable, and thus have fields, and some are not. For example, the LSI
+    //! (LowSpeedInternalRC) clock is always 32 kHz, but the MSI (MediumSpeedInternalRC) clock can
+    //! be configured and thus has a frequency component.
+    //!
+    //! To use them, compose them and feed them to, e.g., sysclk.
+    //!
+    //! ```rust
+    //! let mut rcc = RCC.constrain();
+    //! let msi_clk = clocking::MediumSpeedInternalRC::new(8_000_000, false);
+    //! let sys_clk_src = clocking::SysClkSource::MSI(msi_clk);
+    //! let cfgr = rcc.cfgr.sysclk(sys_clk_src);
+    //! ```
+    //!
+    //! The PLL is a bit more complex because it _is_ a source (`PLLClkOutput`) and also _requires_
+    //! a source (`PLLClkSource`), but you compose the types similarly.
+
+    use super::rcc;
+
+    pub trait InputClock {
+        fn freq(&self) -> u32;
+    }
+
+    /// High-speed internal 16 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedInternal16RC {
+        pub always_on: bool,
+        pub auto_start: bool,
+    }
+
+    impl HighSpeedInternal16RC {
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| w
+                          .hsion()
+                          .set_bit()
+                          .hsikeron()
+                          .bit(self.always_on)
+                          .hsiasfs()
+                          .bit(self.auto_start)
+                         );
+            while rcc.cr.read().hsirdy().bit_is_clear() {}
+            (16_000_000, 0b01)
+        }
+    }
+
+    /// Medium-speed internal 100 kHz - 48 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct MediumSpeedInternalRC {
+        f: u32,
+        auto_cal: bool,
+        bits: u8,
+    }
+
+    impl MediumSpeedInternalRC {
+
+        pub fn new(f: u32, a: bool) -> Self {
+            let bits = match f {
+                100_000 =>    0b0000,
```

Comment:
Can you please add reference to manual chapter from where these values come from?